osu.Game/Database/UserLookupCache.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Framework.Allocation;
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+using osu.Game.Users;
+
+namespace osu.Game.Database
+{
+    public class UserLookupCache : MemoryCachingComponent<int, User>
+    {
+        private readonly ConcurrentBag<int> nextTaskIDs = new ConcurrentBag<int>();
+
+        [Resolved]
+        private IAPIProvider api { get; set; }
+
+        private readonly object taskAssignmentLock = new object();
+
+        private Task<List<User>> pendingRequest;
+
+        /// <summary>
+        /// Whether <see cref="pendingRequest"/> has already grabbed its IDs.
+        /// </summary>
+        private bool pendingRequestConsumedIDs;
+
+        public Task<User> GetUser(int userId, CancellationToken token = default) => GetAsync(userId, token);
```

Comment:
I think our general convention is to have an `Async` suffix on async methods...?